src/txrequest.h

```diff
@@ -0,0 +1,342 @@
+ *
+ * - COMPLETED (txid, peer) entries representing txids that have been requested from a peer, but they timed out, a
+ *   NOTFOUND message was received for them, or an invalid response was received. They're only kept around to prevent
+ *   downloading them again. If only COMPLETED entries exist for a given txid remain (so no CANDIDATE or REQUESTED
+ *   ones), all of them are deleted (this is an invariant, and maintained by all operations below).
+ *
+ * The following operations are supported on this data structure:
+ *
+ * - ReceivedInv(txid, peer) adds a new CANDIDATE entry, unless one already exists for that (txid, peer) combination
+ *   (whether it's CANDIDATE, REQUESTED, or COMPLETED).
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
+ *
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is succesfully
+ *   added to the mempool, seen in a block, or for whatever reason we no longer care about it.
+ *
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
+
+ * - GetRequestable(peer) does the following:
+ *   - Convert all REQUESTED entries (for all txids/peers) with (exptime <= now) to COMPLETED entries.
+ *   - Requestable txids are selected: txids which have no REQUESTED entry, and for which this peer has the
+ *     lowest-priority CANDIDATE entry among all CANDIDATEs with (reqtime <= now) at the time of the GetRequestable
+ *     call. As the high bit of the priority is whether or not the peer is inbound, this will always prefer an
+ *     outbound peer as long as one is available.
+ *   - The selected txids are sorted in order of announcement and returned (even if multiple were added at the same
+ *     time, or even when the clock went backwards while they were being added).
+ *
+ * - RequestedTx(peer, txid) converts the CANDIDATE entry for the provided peer and txid into a REQUESTED one, with
+ *   exptime set to (now + timeout). It can ONLY be called immediately after GetRequestable was called (for the same
+ *   peer), with only AlreadyHaveTx and other RequestedTx calls (both for other txids) in between. Any other
+ *   non-const operation removes the ability to call RequestedTx.
+ */
+class TxRequestTracker {
+    //! Configuration option: delay before downloading from inbound connection.
+    const std::chrono::microseconds m_inbound_delay;
+    //! Configuration option: timeout after requesting a download.
+    const std::chrono::microseconds m_timeout;
+
+    //! The various states a (txid,node) pair can be in.
+    //! Note that CANDIDATE is split up into 3 substates (NEW, BEST, OTHER), allowing more efficient implementation.
+    enum class State : uint8_t {
```

Comment:
Is there a reason to specify the underlying type here? Is it required so that it can be packed into 3 bits in the Entry bitfield below?

Reply:
Yeah, my concern is that otherwise an unsigned type may be picked, and I don't know how that would interact with it being used as a bitfield.

Reply:
I'm getting this build warning:

```
In file included from txrequest.cpp:5:0:
./txrequest.h:158:25: warning: ‘TxRequestTracker::Entry::m_state’ is too small to hold all values of ‘enum class TxRequestTracker::State’
         State m_state : 3;
                         ^
```

configure output:

```
  target os     = linux
  build os      = linux-gnu

  CC            = /usr/bin/ccache gcc
  CFLAGS        = -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = /usr/bin/ccache g++ -std=c++11
  CXXFLAGS      =   -fstack-reuse=none -Wstack-protector -fstack-protector-all  -Wall -Wextra -Wformat -Wvla -Wswitch -Wformat-security -Wredundant-decls -Wunused-variable -Wdate-time -Wsign-compare  -Wno-unused-parameter -Wno-implicit-fallthrough   -g -O2 -fno-extended-identifiers
  LDFLAGS       = -pthread  -Wl,-z,relro -Wl,-z,now -pie  
  ARFLAGS       = cr
```

gcc version:

```
→ gcc --version
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
```

Reply:
Looks like that warning is kind-of a bug in gcc (you can apparently put any value from the underlying type in the enum, and it's complaining that they won't all fit), but it's not fixed until gcc 8.4 or 9.3 -- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61414

Reply:
@jnewbery Does it help if you remove the `: uint8_t`?

If this is too annoying we can just store the state as a non-`enum` type.

Reply:
Done. Changed to store it as a `uint8_t`.